core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java

```diff
@@ -52,17 +52,43 @@
     private final DictionaryId dictionarySourceId;
     private final boolean mayHaveNull;
 
-    public DictionaryBlock(int positionCount, Block dictionary, int[] ids)
+    public static Block create(int positionCount, Block dictionary, int[] ids)
     {
-        this(0, positionCount, dictionary, ids, false, false, randomDictionaryId());
+        return createInternal(positionCount, dictionary, ids, randomDictionaryId());
     }
 
     /**
      * This should not only be used when creating a projection of another dictionary block.
      */
-    public DictionaryBlock(int positionCount, Block dictionary, int[] ids, DictionaryId dictionarySourceId)
+    public static Block createProjectedDictionaryBlock(int positionCount, Block dictionary, int[] ids, DictionaryId dictionarySourceId)
     {
-        this(0, positionCount, dictionary, ids, false, false, dictionarySourceId);
+        return createInternal(positionCount, dictionary, ids, dictionarySourceId);
+    }
+
+    private static Block createInternal(int positionCount, Block dictionary, int[] ids, DictionaryId dictionarySourceId)
+    {
+        if (positionCount == 0) {
+            return dictionary.copyRegion(0, 0);
+        }
+        if (positionCount == 1) {
+            return dictionary.getRegion(ids[0], 1);
+        }
+
+        // if dictionary is an RLE then this can just be a new RLE
+        if (dictionary instanceof RunLengthEncodedBlock rle) {
+            return new RunLengthEncodedBlock(rle.getValue(), positionCount);
+        }
+
+        // unwrap dictionary in dictionary
```

Comment:
This is not a correct unwrap as you cannot preserve `dictionarySourceId` now.

Take a look at:
```
topDictBlock1        topDictBlock2
sourceId=A           sourceId=A

nestedDictBlock1     nestedDictBlock2
sourceId=1           sourceId=2
```
(such situation happens at join)

You cannot unwrap it to:
```
unwrappedDictBlock1 unwrappedDictBlock2
sourceId=A          sourceId=A
```

You should assign a new sourceId if unwrapping is done implicitly by `Dictionary` constructor

Reply:
How can `topDictBlock1` and `topDictBlock2` have the same `sourceId` if the underlying dictionaries are different?

Reply:
> How can topDictBlock1 and topDictBlock2 have the same sourceId if the underlying dictionaries are different?

same `sourceId` implies same `ids`, but it’s stronger than that. Two DictionaryBlocks` might have same ids coincidentally, but different `sourceIds`
If you have columns of
```
page=[
  dictA(source: S, ids: X, dict: nestedA),
  dictB(source: S, ids: X, dict: nestedB),
  dictC(source: S, ids: X, dict: nestedC)]
```
then you essentially process it like:
```
page=MultiChannelDict(
  source:S,
  ids:X
  dict=[nestedA, nestedB, nestedC])
```
I hope this analogy makes it clearer